### PR TITLE
Expose Rate Limit Headers

### PR DESCRIPTION
**Pull request by stephenc**
Exposes the rate limit header responses so that consumers of the API can proactively tune their usage.

@reviewbybees

### CODE REVIEW

**Comment by KostyaSha** [comment]
I guess you far away from good gh integration, so soon you will get https://github.com/kohsuke/github-api/issues/303

**Comment by stephenc** [comment]
@kohsuke ping

File: src/main/java/org/kohsuke/github/GitHub.java [hunk 1]
```diff
@@ -287,17 +295,61 @@ public void setConnector(HttpConnector connector) {
...
             // GitHub Enterprise doesn't have the rate limit, so in that case
             // return some big number that's not too big.
             // see issue #78
             GHRateLimit r = new GHRateLimit();
             r.limit = r.remaining = 1000000;
-            long hours = 1000L * 60 * 60;
-            r.reset = new Date(System.currentTimeMillis() + 1 * hours );
-            return r;
+            long hour = 60L * 60L; // this is madness, storing the date as seconds in a Date object
+            r.reset = new Date((System.currentTimeMillis() + hour) / 1000L );
```
Comment (xhumanoid) [inline src/main/java/org/kohsuke/github/GitHub.java]: I am sorry, but what the crazy calculation do you do here?

hour - value in second
System.currentTimeMillis() - return millisecond

(millisecond + second)/ 1000L - strange unpredictable type of second, we can say this type of seconds as X.

By javadoc:
public Date(long date)
date   the milliseconds since January 1, 1970, 00:00:00 GMT.

if we will use new Date( X ), then we can't predict what type of date we will receive.

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: Ha! Read the comment in GHRateLimit... for crazy reasons that class stores the date in Date with seconds expressed as milliseconds.

Without this fixup the reset date will be incorrect as returned by `GHRateLimit.getResetDate()`

I have comments indicating that it be madness... but actually is the fix if least harm

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: Ack the adding of an hours seconds should be outside of the division, but this will still give a result in the future for GHE and as GHE does not have a rate limit...

Comment (xhumanoid) [inline src/main/java/org/kohsuke/github/GitHub.java]: I read about "this is madness, storing the date as seconds in a Date object"
but very confused by the division.

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: https://github.com/kohsuke/github-api/blob/4ee3086b6dcb46913f1341b2500a4790db5813d6/src/main/java/org/kohsuke/github/GHRateLimit.java#L31 is why we need to do the division... and that is madness... but fixing that madness is orthogonal to this enhancement

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: https://github.com/kohsuke/github-api/pull/329 will fix the algebra... but it is not strictly required as a limit of no limit expiring in 3.6 seconds or expiring in 1 h is still a limit of no limit

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: But without correcting for the 1000 fold factor then GHE would say it had a limit of 1000000 being refreshed on Tue Jun 20 01:41:44 IST 48997 which allows 21 API calls per year... so code that uses the rate limit to pre-throttle will run unnecessarily slow on GHE

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: With #329 code running against GHE and proactively throttling will make 277 API requests per second, while without #329 it might try and make 277777 requests per second as its upper limit... which is probably unatainable

Comment (stephenc) [inline src/main/java/org/kohsuke/github/GitHub.java]: Or were you accessing the public field directly (oh how I love the way people make fields public in Java and make migration very difficult without breaking backwards compatibility)

Comment (xhumanoid) [inline src/main/java/org/kohsuke/github/GitHub.java]: >> is why we need to do the division... and that is madness... but fixing that madness is orthogonal to this enhancement

Yes, I agree with this, fixing this madness must doing not in this issue.
